### DISCUSSION TITLE
[2.0][Type Hints] Session Type Annotations incorrect in PyCharm

### DISCUSSION THREAD

**Discussion by erikwrede** [sqlalchemy/sqlalchemy#8741] [Usage Questions]
### Describe the bug

Type Annotations for several methods of `Session` don't work in PyCharm. The Type `_EntityBindKey` is incorrectly resolved to `_O` instead of the actual Model type for the return type annotation `Optional[_O]`:
<img width="408" alt="image" src="https://user-images.githubusercontent.com/9725143/198872465-dea37c66-0ec7-4c13-846a-e19f4d2c3b40.png">
Another variant observed:
<img width="502" alt="image" src="https://user-images.githubusercontent.com/9725143/198873683-eda18d6b-4753-4e81-8cab-6c4b967b29c7.png">

Type hints on the return of the query don't work unless the variable is manually typed, like in 1.4.

The expected behavior is this (DBTurbine is an aliased import, please excuse the confusion): 
<img width="452" alt="image" src="https://user-images.githubusercontent.com/9725143/198872807-a1ac1a41-08e6-4728-a77a-6572d4b8246a.png">

The issue seems to be with the `_EntityBindKey` Type. If i change the signature to use the value of that Type, it works as shown in expected behavior:
```python
# For Reference:
_EntityBindKey = Union[Type[_O], "Mapper[_O]"]
#AsyncSession.get
    async def get(
        self,
        entity: Union[Type[_O], "Mapper[_O]"],#_EntityBindKey[_O],
        ident: _PKIdentityArgument,
        *,
        options: Optional[Sequence[ORMOption]] = None,
        populate_existing: bool = False,
        with_for_update: Optional[ForUpdateArg] = None,
        identity_token: Optional[Any] = None,
        execution_options: _ExecuteOptionsParameter = util.EMPTY_DICT,
    ) -> Optional[_O]:
```

I'm fully aware that this might be a PyCharm bug; but further investigation is definitely worth it to find a workaround.

### To Reproduce

```python
# Have a Model:
class Turbine(Base):
    __tablename__ = "turbines"
    id = Column(Integer(), primary_key=True)
    model = Column(String(280))

# Get your AsyncSession
async def resolve_turbine(id: int) -> Turbine:
    db : AsyncSession = my_sessionmaker()
    turbine = db.get(Turbine,id) # Turbine is not assigned type turbine in the IDE
    turbine.model # autocomplete does not work here
```

### Versions

- Python: 3.10
- SQLAlchemy: 2.0b2
- PyCharm: latest

**Comment by CaselIT** [comment]
Hi,

This definitely seems like a pycharm issue.
The type that causes problems is already an alias for your proposed type `_EntityBindKey = Union[Type[_O], "Mapper[_O]"]`, so this seems pycharm that has kssues resolving the typos something along these lines.

**Comment by erikwrede** [comment]
Thanks for the quick reply!
I just rebuilt my indices and now I get this, which is better but still incorrect (should be `Turbine` not `Type[Turbine]`)
<img width="502" alt="image" src="https://user-images.githubusercontent.com/9725143/198873570-9933680c-4119-4a78-ae3d-4ff5d25aab0a.png">
Something seems to be broken. I might raise an Issue in the PyCharm Bug tracker. Feel free to close this to keep your issues tidy :)
Alternatively we can use this to track if other users experience problems and I can post updates from the PyCharm bug tracker. Up to you!

**Comment by CaselIT** [comment]
The fact that the type is not stable in pycharm definitely is strange. 

Let's wait for mike opinion regarding the issue.

**Comment by zzzeek** [comment]
Well the mapping is not typed correctly in any case (it should use mapped_column() and DeclarativeBase for 2.0 types to work completely), so as far as `turbine.model`, that gives me `Any` in pylance as well.     As far as the "db.get" part that's a coroutine, so pylance correctly shows it as `(variable) turbine: Coroutine[Any, Any, Turbine | None]`.    If I add the `await`, then it shows `(variable) turbine: Turbine | None`.

So if reporting to pycharm I'd seek to establish mypy as the baseline.   Let's make the program fully strict typed and illustrate the return types:

```py
from typing import TYPE_CHECKING

from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.asyncio import async_sessionmaker
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import mapped_column, Mapped


class Base(DeclarativeBase):
    pass


class Turbine(Base):
    __tablename__ = "turbines"
    id: Mapped[int] = mapped_column(Integer(), primary_key=True)
    model: Mapped[str] = mapped_column(String(280))


my_sessionmaker: async_sessionmaker[AsyncSession] = async_sessionmaker()

# Get your AsyncSession
async def resolve_turbine(id: int) -> Turbine:
    db: AsyncSession = my_sessionmaker()
    turbine = await db.get(Turbine, id)

    assert turbine is not None

    if TYPE_CHECKING:
        reveal_type(turbine)
        reveal_type(turbine.model)

    return turbine
```

mypy output:

```
test3.py:31: note: Revealed type is "test3.Turbine"
test3.py:32: note: Revealed type is "builtins.str"
Success: no issues found in 1 source file
```

that's what you give to pycharm so they have a baseline behavioral example.

**Comment by CaselIT** [comment]
I think this could be a discussion, since I don't think sqlalchemy is doing anything it shouldn't here

**Comment by macdjord** [comment]
Actually, I believe this is, in fact, SQLA's problem.

Let's take a closer look at that type hint:
```python
        entity: _EntityBindKey[_O],
```

Well, `_EntityBindKey` is an alias:
```python
_EntityBindKey = Union[Type[_O], "Mapper[_O]"]
```

If we substitute that alias in:
```python
        entity: Union[Type[_O], "Mapper[_O]"][_O],
```

Can you now see the problem? A `Union` is not a generic type. You cannot parameterize it. Certainly not when the generics inside it have already been parameterized!

The correct type hint is simply:
```python
        entity: _EntityBindKey,
```
The generics were parameterized when the alias was defined; you don't need to do it again. This is, unfortunately, less discoverable, because the return type type hint is `_O`, and it is no longer immediately visible within the function definition where that type variable is getting its value. The only solution I can see would to simply omit the alias entirely and make the type hint be `Union[Type[_O], "Mapper[_O]"]`.

Note that this pattern - creating an alias to a parameterized generic then re-parameterizing the alias on use - shows up in many places. At least some of them will be more difficult to fix; for instance, `_EntityType` is defined as:
```python
_EntityType = Union[
    Type[_T], "AliasedClass[_T]", "Mapper[_T]", "AliasedInsp[_T]"
]
```
But `Session` uses it like so:
```python
    @overload
    def query(self, _entity: _EntityType[_O]) -> Query[_O]:
        ...
```
With a *different* parameterization typevar.

**Reply by CaselIT** [reply]
Both mypy and pylance don't agree with what you wrote.
The following type checks correctly in both and the revealed type is correct
```py
from typing import Generic, TypeVar, Union, reveal_type

T = TypeVar("T")
U = TypeVar("U")

Foo = Union[list[T], set[T]]

class Bar(Generic[U]):
    x: Foo[U]

class Baz(Bar[str]):
    pass

reveal_type(Baz().x)

```

**Reply by CaselIT** [reply]
Also in the example above if you replace `x: Foo[U]` with `x: Foo` you get this mypy error
```
file.py:9: error: Missing type parameters for generic type "Foo"  [type-arg]
```

So Union absolutely is generic

**Reply by macdjord** [reply]
Huh. I went and checked PEP 0484 'Type Hints', and the section on type aliases contains this example:
```python
from typing import TypeVar, Iterable, Tuple

T = TypeVar('T', int, float, complex)
Vector = Iterable[Tuple[T, T]]

def inproduct(v: Vector[T]) -> T:
    return sum(x*y for x, y in v)
def dilate(v: Vector[T], scale: T) -> Vector[T]:
    return ((x * scale, y * scale) for x, y in v)
vec = []  # type: Vector[float]
```
So apparently I was wrong; the existing type hints are correct. It seems type aliases are only exactly equivalent to the type expression they are defined with as long as parameterized generics are not involved; once that happens, they magically become (unparameterized) generics themselves.

I'm going to have to:
* Report this as a PyCharm bug
* Poke the Python mailing list to see if the semantics of parameterized aliases are actually defined anywhere

**Reply by zzzeek** [reply]
> Actually, I believe this is, in fact, SQLA's problem.
> 
> Let's take a closer look at that type hint:
> 
> ```python
>         entity: _EntityBindKey[_O],
> ```
> 
> Well, `_EntityBindKey` is an alias:
> 
> ```python
> _EntityBindKey = Union[Type[_O], "Mapper[_O]"]
> ```
> 
> If we substitute that alias in:
> 
> ```python
>         entity: Union[Type[_O], "Mapper[_O]"][_O],
> ```
> 
> Can you now see the problem? A `Union` is not a generic type. You cannot parameterize it. Certainly not when the generics inside it have already been parameterized!


those generics are not parameterized, they have TypeVars in them.  you can certainly parametrize a union like that.

> The generics were parameterized when the alias was defined; you don't need to do it again. This is, unfortunately, less discoverable, because the return type type hint is `_O`, and it is no longer immediately visible within the function definition where that type variable is getting its value. The only solution I can see would to simply omit the alias entirely and make the type hint be `Union[Type[_O], "Mapper[_O]"]`.

that's what _EntityBindKey[_O] means.

take this short program:

```py
from typing import Generic, TypeVar, Any, Union

_T = TypeVar("_T", bound=Any)


class A(Generic[_T]):
    pass


class B(Generic[_T]):
    pass


AorB = Union[A[_T], B[_T]]


x1: AorB

```

run it in mypy.  We get:

```
$ mypy test3.py 
test3.py:16: error: Missing type parameters for generic type "AorB"  [type-arg]
Found 1 error in 1 file (checked 1 source file)
```

`AorB` is not complete.  Needs to be qualified:

```py
x1: AorB[int]
```

Even if the `Union[]` has a combination of non-generic and generic types, mypy would want a type annotation for that `AorB` type.

this should make intuitive sense.  what exactly is an `AorB` if it consists of `A[_T]` or `B[_T]`?   `A[_T]` and `B[_T]` are not complete types.

**Reply by Doomerdinger** [reply]
@macdjord Did you end up making a ticket for PyCharm? I have searched for a bit, but did not find anything.

As a note, while this does not follow the latest conventions and I am not sure that it is technically correct, if you do something like the following
```python3
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import mapped_column, Mapped


class Base(DeclarativeBase):
    pass


class Turbine(Base):
    __tablename__ = "turbines"
    id: Mapped[int] = mapped_column(Integer(), primary_key=True)
    model: str = Column(type_=String(280))

reveal_type(Turbine().model)
```
It works as expected with Mypy (` Revealed type is "builtins.str"`) and also gives proper type hints in PyCharm.

Doing something similar with PickleType is also possible, though in both the `Mapped` and not-mapped cases Mypy detects `Any` as the type, though pycharm seems to use the correct types for code completion.

**Reply by CaselIT** [reply]
That will most likely cause a mypy error if you access at the class level `Turbine.model` to do something like `Turbine.model.like('str')`

**Reply by Doomerdinger** [reply]
Ah so it does.
So only if the use cases are basic and you highly value PyCharm completion is it an option.

**Reply by macdjord** [reply]
@Doomerdinger Yes, I did: https://youtrack.jetbrains.com/issue/PY-60614/Incorrect-handling-of-parameterized-generic-type-alias-using-Type
